### PR TITLE
lsp: Add a flag to debounce didChange notifications

### PR DESCRIPTION
**Pull request by mfussenegger**
Would help with cases as reported in https://github.com/neovim/neovim/issues/14087

### CODE REVIEW

**Comment by mjlbach** [comment]
LMK when ready for review. How will this work if the debounce window extends past the timepoints at which a completion trigger is entered (with comp-e for example)?

**Comment by mjlbach** [comment]
Cool, I can review this weekend. I did a once over and it seemed good? I prefer a lot of the cleanups/encapsulation you did.

 I'm just generally worried about the intersection of the diff logic and the debouncing (like, I'm not sure if some of the assumptions I made are going to hold up if we combine multiple diff changes into one message). Also not sure if we should push to get this in 0.5 or aim for 0.5.1.

**Comment by mfussenegger** [comment]
> I can review this weekend. I did a once over and it seemed good? I prefer a lot of the cleanups/encapsulation you did.

Cool, sounds great.


> I'm just generally worried about the intersection of the diff logic and the debouncing (like, I'm not sure if some of the assumptions I made are going to hold up if we combine multiple diff changes into one message).

I also wasn't entirely sure if what I'm doing is safe to do.
In the case of incremental sync it still forces the diff computation immediately and chain them together. `contentChanges` can be a list and the spec makes it sound as if it should work:

```
/**
 * The actual content changes. The content changes describe single state
 * changes to the document. So if there are two content changes c1 (at
 * array index 0) and c2 (at array index 1) for a document in state S then
 * c1 moves the document from S to S' and c2 from S' to S''. So c1 is
 * computed on the state S and c2 is computed on the state S'.
 *
 * To mirror the content of a document using change events use the following
 * approach:
 * - start with the same initial content
 * - apply the 'textDocument/didChange' notifications in the order you
 *   receive them.
 * - apply the `TextDocumentContentChangeEvent`s in a single notification
 *   in the order you receive them.
 */
contentChanges: TextDocumentContentChangeEvent[];
```


>  Also not sure if we should push to get this in 0.5 or aim for 0.5.1.

I'm also not sure. For me both is fine.

**Comment by mjlbach** [comment]
In the future, it might be a good idea to have a group we could tag like @neovim/testers with volunteers to try running PRs for awhile. I almost missed a couple bugs (and did miss 1) with incremental sync which had to get patched.

 Maybe we can bring in some people that commented on the incremental sync PR to try this out for a week and report back.

**Comment by mfussenegger** [comment]
> In the future, it might be a good idea to have a group we could tag like @neovim/testers with volunteers to try running PRs for awhile.

I think depending on what the PR touches you may have different groups of people who are interested in it. Not sure if this would work out. But maybe something that could be brought up and discussed in discourse?

I've had another idea about the debounce behavior for incremental sync. Instead of computing the changes eagerly, it could keep track of the start and endlines and then de-duplicate overlapping regions and then compute the diff based on that. I think that would be even faster and may be safer to do. What do you think?



Regarding 0.5.0 vs. 0.5.1 - I think I'm now tending more towards targeting this for 0.5.1

**Comment by mjlbach** [comment]
> I've had another idea about the debounce behavior for incremental sync. Instead of computing the changes eagerly, it could keep track of the start and endlines and then de-duplicate overlapping regions and then compute the diff based on that. I think that would be even faster and may be safer to do. What do you think?

This seems like a good idea, but I would want to test it substantially to make sure the content ranges still worked.

> Regarding 0.5.0 vs. 0.5.1 - I think I'm now tending more towards targeting this for 0.5.1

Agreed

**Comment by mfussenegger** [comment]
From my side this is now good to go.

The debounce flag defaults to `nil`, which means "don't debounce at all" so without an explicit config changes this doesn't change the behavior at all. So would be okay to merge already for 0.5

File: runtime/lua/vim/lsp.lua
Comment (mjlbach) [inline runtime/lua/vim/lsp.lua:236]: Why are we checking if config.flags is not nil?

File: runtime/lua/vim/lsp.lua
Comment (mjlbach) [inline runtime/lua/vim/lsp.lua:238]: I'm not sure how I feel about debounce_did_change_notify, I know what it is, but maybe debounce_text_changes? Idk this is nitpicky

File: runtime/lua/vim/lsp.lua
Comment (mjlbach) [inline runtime/lua/vim/lsp.lua:319]: Is there a reason you don't test this in the inner blocK?

**Comment by mjlbach** [comment]
I just looked over everything and it all LGTM, could probably clean up changetracking.prepare a bit more (that's not on you though, because a lot of it is left over from the previous organization

File: runtime/lua/vim/lsp.lua
Comment (mfussenegger) [reply runtime/lua/vim/lsp.lua:236]: `validate_client_config` is called before we do `config.flags = config.flags or {}`, so at this point `conflig.flags` can still be `nil` which is allowed here - the assert should only trigger if flags are present and there is a non numeric `debounce_did_change_notify` property

File: runtime/lua/vim/lsp.lua
Comment (mfussenegger) [reply runtime/lua/vim/lsp.lua:238]: `debounce_text_changes` also sounds good to me :+1: 
I also wasn't very convinced about `debounce_did_change_notify`

File: runtime/lua/vim/lsp.lua
Comment (mfussenegger) [reply runtime/lua/vim/lsp.lua:319]: You mean in the `state.buffers` block?

If so yes: If it's debounced full-text sync then `state.buffers` wouldn't exist but the timer should be resetted anyways.

File: runtime/lua/vim/lsp.lua
Comment (mfussenegger) [reply runtime/lua/vim/lsp.lua]: Pushed the change.

**Review by mjlbach — APPROVED** [review]

**Comment by mjlbach** [comment]
Should probably also squash the rename commit. I just tagged TJ for review but I'm cool merging this. I think if people don't use debounce it should be the same as before.

**Comment by bluz71** [comment]
Any thought of making this the default (after stabilisation)? My feeling is yes, for similar reasons to why incremental sync was made the default, that being efficiency.

For example, [this post](https://github.com/neovim/neovim/issues/13049#issuecomment-783056229) highlights LSC (with a 500ms debounce) vs Neovim LSP with no debounce doing a simple `ciw`; there were just 4 `didChange` notifications from LSC vs 44 for Neovim.

LSC uses a 500ms debounce as [noted here](https://github.com/natebosch/vim-lsc/blob/master/doc/lsc.txt#L218). I was a LSC for nearly two years, it is a mature LSP client that works great. The debounce was never an issue from my perspective.

But I would think 250ms as [suggested by @mfussenegger](https://github.com/neovim/neovim/issues/14087#issuecomment-797686350) would likely be safer and less sleepy (for want of a better term).

I am really looking forward to this, I think it will be very worthwhile.

Best regards.

**Comment by mjlbach** [comment]
It would depend, I would want to merge this and get feedback from users to catch any corner cases. We made incremental sync the default, but there were issues dealing with utf-16 ranges that I just patched (today), so it's definitely not something which would happen before 0.5. 

This PR is mostly waiting on review from TJ, I've already approved it, but it is on the bigger side of client changes (when enabled).

**Comment by bluz71** [comment]
> so it's definitely not something which would happen before 0.5

Agreed. Operative word being stabilisation. But once it does settle down, I genuinely believe this will be worth enabling by default.

0.5.1 was more my thinking or even 0.5.2.

Maybe introduce `debounce_did_change_notify` in 0.5.1 behind a flag, and then enable it by default in 0.5.2 after folks like me live with it for a few weeks or months.

**Comment by mfussenegger** [comment]
Rebased this to include the utf-16/utf-8 change.

@tjdevries could you have a look at this?


I think we could use the work in this PR also to introduce some kind of LSP specific `atomic_change_notification(func)` which could be used in  https://github.com/neovim/neovim/pull/13703

That text edits are done using buf_set_lines is still one major source of extmark hiccups - e.g. https://github.com/neovim/neovim/issues/14307

**Comment by mjlbach** [comment]
And also https://github.com/neovim/neovim/issues/14093 for why we can't use on_bytes (yet)

**Pull request merged**
